webapp/src/ts/services/db-sync.service.ts

```diff
@@ -250,3 +270,93 @@ export class DBSyncService {
     return this.syncMedic(force);
   }
 }
+
+class DbSyncTelemetry {
+  private readonly telemetryKeyword = 'replication';
+  private readonly offlineErrorMessage = 'Failed to fetch';
+  private readonly database;
+  private readonly direction;
+  private readonly start;
+  private readonly lastReplicated;
+  private readonly key;
+  private end;
+
+  constructor(
+    public telemetryService:TelemetryService,
+    database,
+    direction,
+    lastReplicated?,
+  ) {
+    this.database = database;
+    this.direction = direction;
+    this.start = Date.now();
+    this.lastReplicated = lastReplicated;
+    this.key = `${this.telemetryKeyword}:${this.database}:${this.direction}`;
+  }
+
+  private getSuccessKey() {
+    return `${this.key}:success`;
+  }
+
+  private getFailureKey() {
+    return `${this.key}:failure`;
+  }
+
+  private getDocsKey() {
+    return `${this.key}:docs`;
+  }
+
+  private getErrorKey() {
+    return `${this.getFailureKey()}:reason:error`;
+  }
+
+  private getDeniedKey() {
+    return `${this.key}:denied`;
+  }
+
+  private getOfflineKey(service) {
+    return `${this.getFailureKey()}:reason:offline:${service}`;
+  }
+
+  private getLastReplicatedKey() {
+    return `${this.key}:ms-since-last-replicated-date`;
+  }
+
+  private async record(key) {
+    this.end = Date.now();
+    await this.telemetryService.record(key, this.end - this.start);
+    if (this.lastReplicated) {
+      await this.telemetryService.record(this.getLastReplicatedKey(), this.start - this.lastReplicated);
+    }
+  }
+
+  private recordInfo(info?) {
+    if (info) {
+      const nbrDocs = info.push ?
+        (info.push?.docs_read + info.pull?.docs_read) : // result of .sync contains info for push and pull
+        info.docs_read; // result of .replicate contains info for just one direction
+      return this.telemetryService.record(this.getDocsKey(), nbrDocs);
+    }
+  }
+
+  async recordSuccess(info?) {
+    await this.record(this.getSuccessKey());
+    await this.recordInfo(info);
+  }
+
+  async recordFailure(error, knownOnlineState) {
+    await this.record(this.getFailureKey());
+    await this.recordInfo(error?.result);
```

Comment:
If `error` is not null (don't have the `?` modifier in the method definition) `error.result` should work without the `?` operator.

Reply:
The point is we don't know if the error is or is not null.

Reply:
If `error` can be `null | undefined`, should be declared as `error?` in the function definition, but now checking Typescript docs I see that strict null checks are not enforced by default so you can still have that situation (I guess it is because backward compatibility with JS, and the enforcement can be configured in the build settings but it would mess many lines of our codes at this point).

Reply:
Even if we do error? in the function definition, we still need to check before accessing a property, right?